### DISCUSSION TITLE
Support for QUIC + HTTP/3

### DISCUSSION THREAD

**Discussion by colmexdev** [rack/rack#2039] [Q&A]
Hi there!

   While testing HTTP/3 + QUIC with a patched version of Nginx in a development server, I realized there are some things that break web apps.

   Checking logs from Nginx, it seems that it handles the protocol pretty good; checking logs and output from Rails app reveals that HTTP/3 isn't handled properly e.g.

#### Entry in Nginx access log:
  `<client_ip> - - [<datetime>] "GET / HTTP/3.0" 200 <bytes> "<referer>" <client>`

#### Values taken from _request.env_ on Rails app
```
{ ...,
"SERVER_NAME"=>"127.0.0.1", 
"SERVER_PORT"=>"3000", 
"SERVER_SOFTWARE"=>"nginx/1.23.4 Phusion_Passenger/6.0.17", 
"SERVER_PROTOCOL"=>"HTTP/1.1", ... }
```

So I think something's not working properly on request processing: Protocol and hostname are taken as localhost and HTTP1.1 from second request onwards. Due to that, some internal links (e.g. Active Storage redirection links to blobs) and asset urls (e.g. Propshaft links to static assets) break down because host is taken as "127.0.0.1:3000". 

   I'd like to note that I'm not a Ruby nor Rails expert, so I don't know whether this problem must be discussed here or in some other part (perhaps Passenger or Rails). Any feedback given will be pretty much appreciated, although I know it can be too early to talk about QUIC - HTTP/3 support.

   Thanks in advance!

**Comment by ioquatix** [comment]
My guess is nginx only supports an HTTP/1.1 proxy backend and it's probably not configured with the correct hostname. Can you share the entire Nginx configuration?

**Reply by colmexdev** [reply]
Sure, this is a minimal sample of my Nginx configuration:

```
user www-data;
worker_processes auto;
pid /run/nginx.pid;
include /etc/nginx/modules-enabled/*.conf;

events {
        worker_connections 1024;
}

http {
        # Basic Settings
        sendfile on;
        tcp_nopush on;
        tcp_nodelay on;
        keepalive_timeout 65s;
        types_hash_max_size 2048;

        include /etc/nginx/mime.types;
        default_type application/octet-stream;

        log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_user_agent" "$http_x_forwarded_for"';

        # Pagespeed
        pagespeed unplugged;
        pagespeed FileCachePath /var/ngx_pagespeed_cache;
        pagespeed FetchWithGzip off;
        pagespeed HttpCacheCompressionLevel 0;

        # SSL Settings
        ssl_protocols TLSv1.2 TLSv1.3; # Dropping SSLv3, ref: POODLE
        ssl_prefer_server_ciphers on;

        # Brotli
        brotli on;
        brotli_comp_level 6;
        brotli_static on;

        include /etc/nginx/conf.d/*.conf;
}

server {
        listen 443 http3 reuseport;
        listen 443 ssl http2;

        server_name _;
        root /path/to/rails/app;
        ssl_certificate /path/to/cert;
        ssl_certificate_key /path/to/key;

        passenger_enabled on;
        passenger_app_env production;
        passenger_intercept_errors on;
        error_page 500 /500.html;
        error_page 404 /404.html;

        location /cable {
                passenger_app_group_name www_websocket;
                passenger_force_max_concurrent_requests_per_process 0;
        }

        client_max_body_size 100m;

        quic_retry on;
        http3_push off;
        http3_push_preload on;

        location / {
                add_header Alt-Svc 'h3=":$server_port"; ma=86400';
        }

        location ~ ^/(assets|packs) {
                expires max;
                brotli_static on;
        }
}
```

Hope this helps!

**Reply by ioquatix** [reply]
Right, so you are using passenger, my guess is Passenger does not support native HTTP/3 - and the request is being proxied as an HTTP/1.1 request. This is probably the expected behaviour.

**Comment by Raidm404** [comment]
The solution is here  https://github.com/phusion/passenger/issues/2522

**Comment by developertemur** [comment]
#ATTENTION!
#
#DO NOT MODIFY THIS FILE BECAUSE IT WAS GENERATED AUTOMATICALLY,
#SO ALL YOUR CHANGES WILL BE LOST THE NEXT TIME THE FILE IS GENERATED.
server {
	listen 62.209.128.120:443 quic;
	add_header Alt-Svc 'h3=":443"; ma=86400';
	listen 62.209.128.120:443 ssl;
	http2 on;

	ssl_certificate             /usr/local/psa/var/certificates/scfbP2bY1;
	ssl_certificate_key         /usr/local/psa/var/certificates/scfbP2bY1;

	server_name www.intuitsiya.uz;

	location / {
		return 301 https://intuitsiya.uz$request_uri;
	}

	# mailconfig
	location ~* ^/autodiscover/autodiscover\.xml$ {
		try_files $uri @mad;
	}
	location ~* ^(/\.well-known/autoconfig)?/mail/config\-v1\.1\.xml$ {
		try_files $uri @mad;
	}

	location  ~* ^/email\.mobileconfig$ {
		try_files $uri @mad;
	}

	location @mad {
		rewrite ^(.*)$ /mailconfig/ break;

		proxy_pass                          http://127.0.0.1:8880;
		proxy_set_header X-Host             $host;
		proxy_set_header X-Request-URI      $request_uri;
	}
	# mailconfig
}

server {
	listen 62.209.128.120:443 quic;
	add_header Alt-Svc 'h3=":443"; ma=86400';
	listen 62.209.128.120:443 ssl;
#	http2 on;
	http3 on;

	server_name intuitsiya.uz;
	server_name ipv4.intuitsiya.uz;

	ssl_certificate             /usr/local/psa/var/certificates/scfbP2bY1;
	ssl_certificate_key         /usr/local/psa/var/certificates/scfbP2bY1;

	client_max_body_size 134217728;

	# mailconfig
	location ~* ^/autodiscover/autodiscover\.xml$ {
		try_files $uri @mad;
	}
	location ~* ^(/\.well-known/autoconfig)?/mail/config\-v1\.1\.xml$ {
		try_files $uri @mad;
	}

	location  ~* ^/email\.mobileconfig$ {
		try_files $uri @mad;
	}

	location @mad {
		rewrite ^(.*)$ /mailconfig/ break;

		proxy_pass                          http://127.0.0.1:8880;
		proxy_set_header X-Host             $host;
		proxy_set_header X-Request-URI      $request_uri;
	}
	# mailconfig

	access_log "/var/www/vhosts/system/intuitsiya.uz/logs/proxy_access_ssl_log";
	error_log "/var/www/vhosts/system/intuitsiya.uz/logs/proxy_error_log";

	root "/var/www/vhosts/intuitsiya.uz/httpdocs";

	#extension performance-booster begin
	# Additional directives added by Plesk Optimization Settings

	# Additional directives added by Plesk Optimization Settings
	#extension performance-booster end

	#extension sslit begin

	#extension sslit end

	#extension letsencrypt begin
	location ^~ /.well-known/acme-challenge/ {
		root /var/www/vhosts/default/htdocs;

		types { }
		default_type text/plain;

		satisfy any;
		auth_basic off;
		allow all;

		location ~ ^/\.well-known/acme-challenge.*/\. {
			deny all;
		}
	}
	#extension letsencrypt end

	#extension dot-net begin
	location / {
		passenger_enabled           on;
		passenger_app_root          "/var/www/vhosts/intuitsiya.uz/httpdocs";
		passenger_document_root     "/var/www/vhosts/intuitsiya.uz/httpdocs";
		passenger_app_start_command "dotnet /var/www/vhosts/intuitsiya.uz/httpdocs/Intuitsiya.dll --urls \"http://localhost:$PORT\"";
		passenger_user              "intuitsi";
		passenger_group             "psacln";

		passenger_env_var          "ASPNETCORE_ENVIRONMENT" "Development";
		passenger_env_var          "PASSENGER_ENVIRONMENT" "development";
		passenger_env_var          "PASSENGER_POOL_IDLE_TIME" "300";
		passenger_env_var          "environment" "development";
	}

	#extension dot-net end

	#extension security-txt begin
	location = /.well-known/security.txt {
		alias /var/www/vhosts/default/htdocs/.well-known/security.txt;
	}

	#extension security-txt end

	error_page 400 "/error_docs/bad_request.html";
	error_page 401 "/error_docs/unauthorized.html";
	error_page 403 "/error_docs/forbidden.html";
	error_page 404 "/error_docs/not_found.html";
	error_page 500 "/error_docs/internal_server_error.html";
	error_page 405 "/error_docs/method_not_allowed.html";
	error_page 406 "/error_docs/not_acceptable.html";
	error_page 407 "/error_docs/proxy_authentication_required.html";
	error_page 412 "/error_docs/precondition_failed.html";
	error_page 414 "/error_docs/request_uri_too_long.html";
	error_page 415 "/error_docs/unsupported_media_type.html";
	error_page 501 "/error_docs/not_implemented.html";
	error_page 502 "/error_docs/bad_gateway.html";
	error_page 503 "/error_docs/maintenance.html";

	location ^~ /error_docs {
		root "/var/www/vhosts/intuitsiya.uz";
	}

	location ~ /\.ht {
		deny all;
	}

	location ~ ^/(plesk-stat|awstats-icon|webstat|webstat-ssl|ftpstat|anon_ftpstat) {
		auth_basic "Domain statistics";
		auth_basic_user_file "/var/www/vhosts/system/intuitsiya.uz/pd/d..httpdocs@plesk-stat";
		autoindex on;

		location ~ ^/plesk-stat(.*) {
			alias /var/www/vhosts/system/intuitsiya.uz/statistics/$1;
		}

		location ~ ^/awstats-icon(.*) {
			alias /usr/share/awstats/wwwroot/icon/$1;
		}

		location ~ ^/(.*) {
			alias /var/www/vhosts/system/intuitsiya.uz/statistics/$1;
		}
	}

	location ~ ^/~(.+?)(/.*?\.php)(/.*)?$ {
		alias /var/www/vhosts/intuitsiya.uz/web_users/$1/$2;
		fastcgi_split_path_info ^((?U).+\.php)(/?.+)$;
		try_files $uri $fastcgi_script_name =404;
		fastcgi_param PATH_INFO $fastcgi_path_info;
		fastcgi_pass "unix:/var/www/vhosts/system/intuitsiya.uz/php-fpm.sock";
		include /etc/nginx/fastcgi.conf;

	}

	location ~ \.php(/.*)?$ {
		fastcgi_split_path_info ^((?U).+\.php)(/?.+)$;
		try_files $uri $fastcgi_script_name =404;
		fastcgi_param PATH_INFO $fastcgi_path_info;
		fastcgi_pass "unix:/var/www/vhosts/system/intuitsiya.uz/php-fpm.sock";
		include /etc/nginx/fastcgi.conf;

	}

	index "index.html" "index.cgi" "index.pl" "index.php" "index.xhtml" "index.htm" "index.shtml";

	disable_symlinks if_not_owner "from=/var/www/vhosts/intuitsiya.uz";

	add_header X-Powered-By PleskLin;

}
server {
	listen 62.209.128.120:80;

	server_name www.intuitsiya.uz;

	location / {
		return 301 https://intuitsiya.uz$request_uri;
	}

	# mailconfig
	location ~* ^/autodiscover/autodiscover\.xml$ {
		try_files $uri @mad;
	}
	location ~* ^(/\.well-known/autoconfig)?/mail/config\-v1\.1\.xml$ {
		try_files $uri @mad;
	}

	location  ~* ^/email\.mobileconfig$ {
		try_files $uri @mad;
	}

	location @mad {
		rewrite ^(.*)$ /mailconfig/ break;

		proxy_pass                          http://127.0.0.1:8880;
		proxy_set_header X-Host             $host;
		proxy_set_header X-Request-URI      $request_uri;
	}
	# mailconfig
}

server {
	listen 62.209.128.120:80;

	server_name intuitsiya.uz;
	server_name ipv4.intuitsiya.uz;

	client_max_body_size 134217728;

	# mailconfig
	location ~* ^/autodiscover/autodiscover\.xml$ {
		try_files $uri @mad;
	}
	location ~* ^(/\.well-known/autoconfig)?/mail/config\-v1\.1\.xml$ {
		try_files $uri @mad;
	}

	location  ~* ^/email\.mobileconfig$ {
		try_files $uri @mad;
	}

	location @mad {
		rewrite ^(.*)$ /mailconfig/ break;

		proxy_pass                          http://127.0.0.1:8880;
		proxy_set_header X-Host             $host;
		proxy_set_header X-Request-URI      $request_uri;
	}
	# mailconfig

	access_log "/var/www/vhosts/system/intuitsiya.uz/logs/proxy_access_log";
	error_log "/var/www/vhosts/system/intuitsiya.uz/logs/proxy_error_log";

	location / {
		return 301 https://$host$request_uri;
	}
}


aspnetcore 8 returns 400 with GET HTTP/3.0 request and don't mismatch HOST Header
